[cocoa] Adjust _allowOnlyPartitionedCookies on HTTP redirect

sysrqb opened the pull request:
#### 66c65ee317db886cfbb7321f244b2a4ab3da4c5a
<pre>
[cocoa] Adjust _allowOnlyPartitionedCookies on HTTP redirect
<a href="https://bugs.webkit.org/show_bug.cgi?id=284660">https://bugs.webkit.org/show_bug.cgi?id=284660</a>
<a href="https://rdar.apple.com/141459915">rdar://141459915</a>

Reviewed by Sihui Liu.

When partitioned cookies are enabled, we need to adjust the
_allowOnlyPartitionedCookies of the request in two places:
1) When the request is initiated (e.g., when we are creating the task)
2) When a redirect is handled

This patch refactors some logic so it can be shared between the
NetworkDataTaskCocoa constructor and
NetworkTaskCocoa::willPerformHTTPRedirection:

1) obtaining the current third party cookies blocking mode
3) deciding if third party cookies should be blocked
4) checking if opt-in cookie partitioning is enabled

* LayoutTests/http/tests/resourceLoadStatistics/only-partitioned-cookies-after-redirect.https-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/only-partitioned-cookies-after-redirect.https.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::NetworkTaskCocoa::shouldBlockCookies):
(WebKit::NetworkTaskCocoa::requestThirdPartyCookieBlockingDecision const):
(WebKit::NetworkTaskCocoa::willPerformHTTPRedirection):

Canonical link: <a href="https://commits.webkit.org/288185@main">https://commits.webkit.org/288185@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8edd977505cd3f94215ca102a89c0766cf82334f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82038 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1565 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35995 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86595 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33070 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1600 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9391 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63950 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21674 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85108 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1186 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74640 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44234 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1088 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28817 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31490 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29429 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88028 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9280 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6613 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72321 "Found 1 new test failure: compositing/animation/repaint-after-clearing-shared-backing.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9465 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70459 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71543 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15662 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14576 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12730 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9231 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14767 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9071 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12597 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10879 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -239,6 +239,26 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
+    return thirdPartyCookieBlockingDecision == WebCore::ThirdPartyCookieBlockingDecision::All;
+}
+
+std::pair<bool, WebCore::ThirdPartyCookieBlockingDecision> NetworkTaskCocoa::requestThirdPartyCookieBlockingDecision(const WebCore::ResourceRequest& request) const
```
It might be clearer to separate `bool isOptInCookiePartitioningEnabled()` (which is only available under `HAVE(ALLOW_ONLY_PARTITIONED_COOKIES)`) and `WebCore::ThirdPartyCookieBlockingDecision requestThirdPartyCookieBlockingDecision`.

Commit:
[cocoa] Adjust _allowOnlyPartitionedCookies on HTTP redirect
https://bugs.webkit.org/show_bug.cgi?id=284660
rdar://141459915

Reviewed by Sihui Liu.

When partitioned cookies are enabled, we need to adjust the
_allowOnlyPartitionedCookies of the request in two places:
1) When the request is initiated (e.g., when we are creating the task)
2) When a redirect is handled

This patch refactors some logic so it can be shared between the
NetworkDataTaskCocoa constructor and
NetworkTaskCocoa::willPerformHTTPRedirection:

1) obtaining the current third party cookies blocking mode
3) deciding if third party cookies should be blocked
4) checking if opt-in cookie partitioning is enabled

* LayoutTests/http/tests/resourceLoadStatistics/only-partitioned-cookies-after-redirect.https-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/only-partitioned-cookies-after-redirect.https.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::NetworkTaskCocoa::shouldBlockCookies):
(WebKit::NetworkTaskCocoa::requestThirdPartyCookieBlockingDecision const):
(WebKit::NetworkTaskCocoa::willPerformHTTPRedirection):

Canonical link: https://commits.webkit.org/288185@main
